RadixWallet/Clients/CloudBackupClient/CloudBackupClient+Live.swift

```diff
@@ -8,81 +8,106 @@ extension CKRecord.RecordType {
-
 	public static func live(
 		profileStore: ProfileStore = .shared
 	) -> CloudBackupClient {
 		@Dependency(\.secureStorageClient) var secureStorageClient
 		@Dependency(\.userDefaults) var userDefaults
 
-		let cloudContainer = (Bundle.main.infoDictionary?[Self.cloudBackupIdentifierKey] as? String).map(CKContainer.init)
-
-		@Sendable
-		func container() throws -> CKContainer {
-			guard let cloudContainer else { throw MissingCloudKitIdentifierError() }
-			return cloudContainer
-		}
+		let container: CKContainer = .default()
 
 		@Sendable
 		func fetchProfileRecord(_ id: CKRecord.ID) async throws -> CKRecord {
-			try await container().privateCloudDatabase.record(for: id)
+			try await container.privateCloudDatabase.record(for: id)
 		}
 
 		@Sendable
-		func fetchAllProfileRecords() async throws -> [CKRecord] {
-			let records = try await container().privateCloudDatabase.records(
-				matching: .init(recordType: .profile, predicate: .init(value: true))
+		func fetchAllProfileRecords(headerOnly: Bool = false) async throws -> [CKRecord] {
+			let records = try await container.privateCloudDatabase.records(
+				matching: .init(recordType: .profile, predicate: .init(value: true)),
+				desiredKeys: headerOnly ? .header : nil
 			)
 			return try records.matchResults.map { try $0.1.get() }
 		}
 
 		@Sendable
-		func extractProfile(_ record: CKRecord) throws -> BackedupProfile {
+		func getProfile(_ record: CKRecord) throws -> BackedUpProfile {
 			guard record.recordType == .profile else {
 				throw IncorrectRecordTypeError()
 			}
 			guard let asset = record[.content] as? CKAsset, let fileURL = asset.fileURL else {
 				throw NoProfileInRecordError()
 			}
 
-			let data = try Data(contentsOf: fileURL)
-			let containsLegacyP2PLinks = Profile.checkIfProfileJsonContainsLegacyP2PLinks(contents: data)
-			let profile = try Profile(jsonData: data)
+			let json = try String(contentsOf: fileURL, encoding: .utf8)
+			let containsLegacyP2PLinks = Profile.checkIfProfileJsonStringContainsLegacyP2PLinks(jsonString: json)
+			let profile = try Profile(jsonString: json)
 			try FileManager.default.removeItem(at: fileURL)
 
-			return BackedupProfile(profile: profile, containsLegacyP2PLinks: containsLegacyP2PLinks)
-		}
+			guard try getProfileHeader(record) == profile.header else {
+				throw HeaderAndMetadataMismatchError()
+			}
 
-		@discardableResult
-		@Sendable
-		func uploadProfileToICloud(_ profile: Profile, existingRecord: CKRecord?) async throws -> CKRecord {
-			try await uploadProfileSnapshotToICloud(profile.profileSnapshot(), id: profile.id, existingRecord: existingRecord)
+			return BackedUpProfile(profile: profile, containsLegacyP2PLinks: containsLegacyP2PLinks)
 		}
 
 		@discardableResult
 		@Sendable
-		func uploadProfileSnapshotToICloud(_ profileSnapshot: Data, id: ProfileID, existingRecord: CKRecord?) async throws -> CKRecord {
+		func uploadProfileToICloud(_ profile: Profile, existingRecord: CKRecord?) async throws -> CKRecord {
```

Comment:
@kugel3 the reason for having previously to upload raw Data instead of Profile, is that decoding the snapshot from the Keychain during migration will delete P2PLinks information, as now Profile does not have it anymore.
So just by the effect of decoding and then encoding, P2PLinks will be lost.

Reply:
But if it's not part of the profile definition anymore, why would we want to keep that information?

Reply:
so we present an alert to the users that Wallet now supports the new CE and  that they need to re link.

Reply:
more precisely see `getProfile` in this file, that it checks for legacy P2PLinks

Reply:
Sure, but do we actually care about this for backups of old profiles, that are not your main profile?

Reply:
In my understanding we should inform users that their previous P2PLinks are lost and they need to re-link whenever the Wallet detects that - be it on the currently profile, on the profile from the cloud backup or on the profile imported from a file.

Additionally, an upload of the Profile to cloud should ideally not loose any information. True, this is a very particular case when the decoded representation of the Profile does not match the underlying bytes, but we should let other parts of the codebase handle it, cloud backups should not make any assumptions on its behalf.

Reply:
> should ideally not loose any information
This should of course be true, but i would say it's very bad practice to rely on information being present in an object, that is no longer part of the definition of that object. Either we care about it, then it should be part of the public contract, or we don't, and then we should make no effort to keep it there, or rely on it being present.

>cloud backups should not make any assumptions on its behalf.
I don't agree with this. I think all parts of the code should be able to rely on the public schema of objects, data structures should not have hidden data.

If we care about this information, it should be kept around as a property in the profile, or alternatively stored in metadata. Those are the robust solutions.

Nevertheless, I refactored it a bit so that it does retain that information.